src/v/config/tls_config.cc

```diff
@@ -150,19 +184,39 @@ bool convert<config::tls_config>::decode(
       ^ static_cast<bool>(node["cert_file"])) {
         return false;
     }
+
+    // Must have either both or neither for PKCS#12 files
+    if (
+      static_cast<bool>(node["p12_file"])
+      ^ static_cast<bool>(node["p12_password"])) {
+        return false;
+    }
+
+    // Cannot have both key/cert and P12 file
+    if (
+      static_cast<bool>(node["key_file"])
+      && static_cast<bool>(node["p12_file"])) {
+        return false;
+    }
     auto enabled = node["enabled"] && node["enabled"].as<bool>();
     if (!enabled) {
         rhs = config::tls_config(
           false, std::nullopt, std::nullopt, std::nullopt, false);
     } else {
-        auto key_cert = node["key_file"] ? std::make_optional<config::key_cert>(
-                          config::key_cert{
-                            to_absolute(node["key_file"].as<ss::sstring>()),
-                            to_absolute(node["cert_file"].as<ss::sstring>())})
-                                         : std::nullopt;
+        std::optional<config::key_cert_container> container;
+        if (node["key_file"]) {
+            container.emplace(config::key_cert{
+              to_absolute(node["key_file"].as<ss::sstring>()),
+              to_absolute(node["cert_file"].as<ss::sstring>())});
+        } else if (node["p12_file"]) {
+            container = std::make_optional<config::key_cert_container>(
```

Comment:
nitpick: presumably could `emplace` this one too, but I don't think it matters at all